ddtrace/tracer/textmap_test.go

```diff
@@ -105,6 +106,100 @@ func TestTextMapCarrierForeachKeyError(t *testing.T) {
+			*/
+			name:                      "datadog-and-w3c",
+			propagationStyle:          "datadog,tracecontext",
+			traceparent:               "00-00000000000000000000000000000004-2222222222222222-01",
+			wantTracestatePropagation: true,
+		},
+		{
+			/*
+				With Datadog AND w3c propagation set, but mismatching trace-ids,
+				the tracestate header should be ignored, and not propagated to
+				the returned trace context.
+			*/
+			name:                      "datadog-and-w3c-mismatching-ids",
+			propagationStyle:          "datadog,tracecontext",
+			traceparent:               "00-00000000000000000000000000000088-2222222222222222-01",
+			wantTracestatePropagation: false,
+		},
+		{
+			/*
+				With Datadog AND w3c propagation set, but mismatching trace-ids,
+				the tracestate header should be ignored, and not propagated to
+				the returned trace context.
+			*/
+			name:                      "datadog-and-w3c-malformed",
+			propagationStyle:          "datadog,tracecontext",
+			traceparent:               "00-00000000000000000000000000000004-22asdf!2-01", // malformed
+			wantTracestatePropagation: false,
+		},
+		{
+			/*
+				With Datadog AND w3c propagation set, but DD_TRACE_PROPAGATION_EXTRACT_FIRST
+				is true, so the tracestate header should be ignored, and not propagated to
+				the returned trace context.
+			*/
+			name:                      "datadog-and-w3c-only-extract-first",
+			propagationStyle:          "datadog,tracecontext",
+			traceparent:               "00-00000000000000000000000000000004-2222222222222222-01", // malformed
+			onlyExtractFirst:          true,
+			wantTracestatePropagation: false,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(fmt.Sprintf("TestTextMapExtractTracestatePropagation-%s", tc.name), func(t *testing.T) {
+			t.Setenv(headerPropagationStyle, tc.propagationStyle)
+			if tc.onlyExtractFirst {
+				os.Setenv("DD_TRACE_PROPAGATION_EXTRACT_FIRST", "true")
+				defer os.Unsetenv("DD_TRACE_PROPAGATION_EXTRACT_FIRST")
+			}
+			tracer := newTracer()
+			assert := assert.New(t)
+			headers := TextMapCarrier(map[string]string{
+				DefaultTraceIDHeader:  "4",
+				DefaultParentIDHeader: "1",
+				traceparentHeader:     tc.traceparent,
+				tracestateHeader:      "dd=s:2;o:rum;t.tid:1230000000000000~~,othervendor=t61rcWkgMzE",
+			})
+
+			ctx, err := tracer.Extract(headers)
+			assert.Nil(err)
+			sctx, ok := ctx.(*spanContext)
+			assert.True(ok)
+			assert.Equal("00000000000000000000000000000004", sctx.traceID.HexEncoded())
+			assert.Equal(uint64(1), sctx.spanID)
+			if tc.wantTracestatePropagation {
+				assert.Equal("dd=s:2;o:rum;t.tid:1230000000000000~~,othervendor=t61rcWkgMzE", sctx.trace.propagatingTag(tracestateHeader))
+			} else if sctx.trace != nil {
+				assert.False(sctx.trace.hasPropagatingTag(tracestateHeader))
+			}
+			// TODO: This could still benefit from some tests verifying that other fields are not changed,
+			// and only the tracestate header was propagated.
```

Comment:
I'd still like to do this before merge, but could use a first review while I work on it.